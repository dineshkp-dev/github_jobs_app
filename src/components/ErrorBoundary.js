import React, { Component } from 'react';
import '../styles/error.css';
import { HOME_PAGE_LINK } from '../common/constants';

export class ErrorBoundary extends Component {
  state = { error: false };

  static getDerivedStateFromError(error) {
    return { error };
  }

  render() {
    const errorMessage =
      this.state.error && this.state.error.message
        ? this.state.error.message
        : 'Oops! An Error has occurred.';
    return this.state.error ? (
      <div className='error-container'>
        <h1 className='error-title'>Error!</h1>
        <div className='error-msg'>
          <div className='error-text'>{errorMessage}</div>
          <div>
            Please click <a href={HOME_PAGE_LINK}>here</a> to go back to the
            home page
          </div>
        </div>
      </div>
    ) : (
      this.props.children
    );
  }
}

export default ErrorBoundary;
