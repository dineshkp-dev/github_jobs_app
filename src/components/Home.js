import React, { useState, useEffect } from 'react';
import Table from './Table';
import { getJobListing } from '../services/jobListingService';
import LoadingIndicator from './LoadingIndicator';
import ListError from './ListError';
import { useAsyncError } from '../common/commonUtils';

const Home = () => {
  const [jobListing, setJobListing] = useState([]);
  const [error, setError] = useState(false);
  const [errorMsg, setErrorMsg] = useState('');

  const throwError = useAsyncError();
  useEffect(() => {
    async function getJobListingFromService() {
      try {
        let listResponse = await getJobListing();
        if (listResponse.error) {
          setError(true);
          setErrorMsg(listResponse.errorMsg);
        } else {
          setJobListing(listResponse.listings);
        }
      } catch (e) {
        throwError(new Error(e));
      }
    }
    getJobListingFromService();
  }, [throwError]);
  return (
    <div className='start'>
      {jobListing && jobListing.length ? (
        <Table jobs={jobListing} />
      ) : error ? (
        <ListError text={errorMsg} />
      ) : (
        <LoadingIndicator text='Please wait for job listings to be updated' />
      )}
    </div>
  );
};

export default Home;
