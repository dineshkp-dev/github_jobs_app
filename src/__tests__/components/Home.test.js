import Home from '../../components/Home';
import * as React from 'react';
import { render, screen, waitFor } from '@testing-library/react';
import { listings } from '../utils/testUtils';
import ErrorBoundary from '../../components/ErrorBoundary';
import { BrowserRouter as Router } from 'react-router-dom';
import '@testing-library/jest-dom/extend-expect';

const jobListingService = require('../../services/jobListingService');

describe('Home snapshot test', () => {
  test('renders', () => {
    jobListingService.getJobListing = jest.fn(() => Promise.resolve(listings));
    const { container } = render(<Home />);
    expect(container).toMatchSnapshot();
  });
});

describe('Home render test', () => {
  test('renders Home', () => {
    jobListingService.getJobListing = jest.fn(() => Promise.resolve(listings));
    render(<Home />);
    expect(screen.getByText('Loading...')).toBeInTheDocument();
    expect(
      screen.getByText('Please wait for job listings to be updated')
    ).toBeInTheDocument();
  });
});

describe('Home get results test', () => {
  test('get success results from Listing Service', async () => {
    jobListingService.getJobListing = jest.fn(() => Promise.resolve(listings));
    render(
      <Router>
        <Home />
      </Router>
    );
    await waitFor(() => {
      screen.getByText('Type');
      screen.getByText('Title');
      screen.getByText('Location');
      screen.getByText('Full Time');
      screen.getByText('Full Stack Engineer');
      screen.getByText('New York City');
      screen.getByPlaceholderText('Filter by job-location');
    });
  });

  test('get error results from Listing Service', async () => {
    let errorListings = listings;
    errorListings.error = true;
    errorListings.errorMsg = 'Dummy error message';
    jobListingService.getJobListing = jest.fn(() =>
      Promise.resolve(errorListings)
    );
    render(<Home />);
    await waitFor(() => {
      screen.getByText('No listing found');
    });
    screen.getByText('Dummy error message');
  });

  test('Reject from Listing Service', async () => {
    jobListingService.getJobListing = jest.fn(() =>
      Promise.reject('DUMMY error')
    );
    render(
      <ErrorBoundary>
        <Home />
      </ErrorBoundary>
    );
    await waitFor(() => {
      screen.getByText('Error!');
      screen.getByText('DUMMY error');
      screen.getByText(/Please click/i);
      screen.debug();
    });
  });

  test('Exception from Listing Service', async () => {
    jobListingService.getJobListing = jest.fn(() => {
      throw new Error('DUMMY error');
    });
    render(
      <ErrorBoundary>
        <Home />
      </ErrorBoundary>
    );
    await waitFor(() => {
      screen.getByText('Error!');
      screen.getByText('Error: DUMMY error');
      screen.getByText(/Please click/i);
      screen.debug();
    });
  });
});
