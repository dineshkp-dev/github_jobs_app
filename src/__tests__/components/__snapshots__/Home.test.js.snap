// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home snapshot test renders 1`] = `
<div>
  <div
    class="start"
  >
    <div
      class="loading-indicator-container"
    >
      <h3
        class="loading-title"
      >
        Loading...
      </h3>
      <div
        class="loading-text"
      >
        Please wait for job listings to be updated
      </div>
    </div>
  </div>
</div>
`;
